File: src/sud.c
// SPDX-License-Identifier: GPL-3.0-only
/*
 *  sud/src/server.c
 *
 *  Copyright (C) Emily <info@emy.sh>
 *  Copyright (C) Kat <kat@castellotti.net>
 */

#include <stdio.h>
#include <sud/args.h>
#include <sud/auth.h>
#include <sud/config.h>
#include <sud/exec.h>
#include <sud/sud.h>

pid_t sud_handle(int conn_fd, int *error) {
    int rc;
    pid_t exec_pid = -1;
    sud_global_config_t global_conf = {};
    process_info_t pinfo;
    user_info_t original_user_info;
    user_info_t target_user_info;
    sud_cmdline_args_t args;
    char *process_path;
    char cmd[PATH_MAX];

    rc = load_global_config(&global_conf);
    if (rc < 0) {
        SUD_ERR("Failed to load global config\n");
        goto exit;
    }

    rc = get_process_info_conn(conn_fd, &pinfo);
    if (rc < 0) {
        SUD_ERR("Failed to get process info from procfs\n");
        goto exit;
    }

    rc = get_userinfo_from_pid(pinfo.uid, &original_user_info);
    if (rc < 0) {
        SUD_ERR("Failed to get original user info\n");
        return -1;
    }

    rc = get_userinfo_from_pid(args.user, &target_user_info);
    if (rc < 0) {
        SUD_ERR("Failed to get original user info\n");
        return -1;
    }

    rc = parse_cmdline_silence(pinfo.argc, pinfo.argv, &args);
    if (rc != 0) {
        SUD_ERR("Cmdline parser error\n");
        goto exit;
    }

    process_path = getenv_envp("PATH", pinfo.envp);
    if (!process_path) {
        SUD_ERR("Missing env PATH in caller process\n");
        goto exit;
    }

    if (!args.workdir) {
        args.workdir = pinfo.cwd;
    }

    if (!args.background_color) {
        args.background_color = global_conf.background_color;
    }

    rc = find_executable(args.argv[0], process_path, args.workdir, cmd);
    if (rc < 0) {
        SUD_ERR("Can't find cmd to be execute'\n");
        goto exit;
    }

    args.cmd = cmd;
    args.argc = args.argc - 1;
    args.argv = &(args.argv[1]);

    SUD_FNOTICE("Authentication for user %d from process %d started\n", pinfo.uid, pinfo.pid);

    if (!sud_auth(&pinfo, &original_user_info, &target_user_info, &args, &global_conf)) {
        SUD_FERR("Authentication for user %d from process %d failed\n", pinfo.uid, pinfo.pid);
        *error = SUD_MSG_ERROR_AUTH;
        goto exit;
    }

    SUD_FNOTICE("Authentication for user %d from process %d completed successfully\n", pinfo.uid, pinfo.pid);

    exec_pid = sud_exec(&pinfo, &original_user_info, &target_user_info, &args);
    if (exec_pid < 0) {
        SUD_ERR("Failed to start exec process\n");
        goto exit;
    }

    SUD_FNOTICE("Executed process %d authenticated as user %d\n", exec_pid, pinfo.uid);

exit:
    free_process_info(&pinfo);
    free_userinfo(&original_user_info);
    free_userinfo(&target_user_info);
    return exec_pid;
}

File: src/sud.rs
// SPDX-License-Identifier: GPL-3.0-only
/*
 *  sud/src/server.rs
 *
 *  Copyright (C) Emily <info@emy.sh>
 *  Copyright (C) Kat <kat@castellotti.net>
 */

use crate::args::SudCmdlineArgs;
use crate::auth::SudAuthPersist;
use crate::auth::{UserInfo, sud_auth};
use crate::config::SudGlobalConfig;
use crate::exec::sud_exec;
use crate::utils::ProcessInfo;
use clap;
use nix;
use std::fmt;
use std::io;
use std::mem;
use std::num;
use std::os::fd::BorrowedFd;
use std::process::Child;
use std::sync::Arc;
use std::sync::Mutex;

pub const SUD_SOCKET_PATH: &str = "sud_privilege_manager_socket";
pub const SUD_MAGIC: &str = "____sud_privilege_manager____";

#[derive(Debug)]
pub enum SudError {
    GenericFuncError(String, String),
    GenericUnsafeCall(String, i32, String),
    InvalidConfig(String),
    NotFound(String),
    AuthFail(String),
    IoError(io::Error),
    ParseIntError(num::ParseIntError),
    NixError(nix::errno::Errno),
    ClapError(clap::error::Error),
}

impl fmt::Display for SudError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SudError::GenericFuncError(fn_name, msg) => {
                write!(f, "Error in fn {}(): {}", fn_name, msg)
            }
            SudError::GenericUnsafeCall(fn_name, code, msg) => write!(
                f,
                "Unsafe call to fn {}() failed with code {}: {}",
                fn_name, code, msg
            ),
            SudError::InvalidConfig(msg) => write!(f, "Invalid config: {}", msg),
            SudError::NotFound(msg) => write!(f, "Not found: {}", msg),
            SudError::AuthFail(msg) => write!(f, "Auth failed: {}", msg),
            SudError::IoError(e) => write!(f, "IO error ({}): {}", e.kind(), e.to_string()),
            SudError::ParseIntError(e) => write!(f, "Parse int error: {}", e.to_string()),
            SudError::NixError(e) => write!(f, "Error in nix function: {}", e.to_string()),
            SudError::ClapError(e) => write!(f, "Error in args parsing: {}", e.to_string()),
        }
    }
}

impl From<io::Error> for SudError {
    fn from(err: io::Error) -> SudError {
        SudError::IoError(err)
    }
}

impl From<num::ParseIntError> for SudError {
    fn from(err: num::ParseIntError) -> SudError {
        SudError::ParseIntError(err)
    }
}

impl From<nix::errno::Errno> for SudError {
    fn from(err: nix::errno::Errno) -> SudError {
        SudError::NixError(err)
    }
}

impl From<clap::error::Error> for SudError {
    fn from(err: clap::error::Error) -> SudError {
        SudError::ClapError(err)
    }
}

#[repr(u32)]
#[derive(PartialEq, Clone, Debug)]
pub enum SudMsgError {
    Success = 0x00,
    Generic = 0x01,
    Auth = 0x02,
    ClientError = 0x03,
}

#[repr(C)]
pub struct SudResponseMsg {
    pub magic: [u8; SUD_MAGIC.len()],
    pub exit_code: i32,
    pub error: SudMsgError,
}

impl Default for SudResponseMsg {
    fn default() -> Self {
        Self {
            magic: SUD_MAGIC.as_bytes().try_into().unwrap(),
            exit_code: 0,
            error: SudMsgError::Generic,
        }
    }
}

impl SudResponseMsg {
    pub fn as_bytes(&self) -> &[u8] {
        unsafe { std::slice::from_raw_parts(self as *const _ as *const u8, mem::size_of::<Self>()) }
    }

    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
        if bytes.len() < mem::size_of::<Self>() {
            return None;
        }

        let msg = unsafe { std::ptr::read(bytes.as_ptr() as *const SudResponseMsg) };

        if msg.magic != SUD_MAGIC.as_bytes() {
            return None;
        }

        Some(msg)
    }
}

pub fn sud_handle(
    conn_fd: BorrowedFd,
    global_config: &SudGlobalConfig,
    auth_persists: Arc<Mutex<Vec<SudAuthPersist>>>,
) -> Result<Child, SudError> {
    let mut pinfo = ProcessInfo::from_conn(conn_fd)?;

    let original_userinfo = UserInfo::from_uid(pinfo.uid)?;

    let args = SudCmdlineArgs::parse(&pinfo)?;

    let target_userinfo = args.get_user()?;

    println!(
        "Authentication for user {} ({}) as target user {} ({}) from process {} trying to execute {} {} started",
        original_userinfo.user.name,
        original_userinfo.user.uid,
        target_userinfo.user.name,
        target_userinfo.user.uid,
        pinfo.pid,
        args.command.as_ref().unwrap(),
        args.args
            .is_empty()
            .then(|| String::new())
            .unwrap_or_else(|| args.args.join(" ")),
    );

    if !sud_auth(
        &mut pinfo,
        &original_userinfo,
        &target_userinfo,
        &args,
        &global_config,
        auth_persists,
    )? {
        return Err(SudError::AuthFail(format!(
            "Authentication for user {} from process {} failed",
            pinfo.uid, pinfo.pid
        )));
    }

    println!(
        "Authentication for user {} from process {} completed successfully",
        pinfo.uid, pinfo.pid
    );

    let child = sud_exec(
        &pinfo,
        &original_userinfo,
        &target_userinfo,
        &args,
        global_config,
    )?;

    println!(
        "Executed process {} ({} {}) as user {} ({}) authenticated as user {} ({})",
        child.id(),
        args.command.unwrap(),
        args.args
            .is_empty()
            .then(|| String::new())
            .unwrap_or_else(|| args.args.join(" ")),
        target_userinfo.user.name,
        target_userinfo.user.uid,
        original_userinfo.user.name,
        original_userinfo.user.uid,
    );

    Ok(child)
}
